base/common/python/pki/nssdb.py

```diff
@@ -1156,6 +1156,24 @@ def __list_private_keys(self, password_file, token=None):
         # The hex string is the hex-encoded CKA_ID
         return re.findall(br'^<\s*\d+>\s+\w+\s+(\w+)', out, re.MULTILINE)
 
+    def __create_basic_constraints_ext(self, exts, basic_constraints_ext):
+        '''
+        Create basic constraints extension config for pki nss-cert-request/issue.
+        '''
+
+        values = []
+
+        if basic_constraints_ext.get('critical'):
+            values.append('critical')
+
+        if basic_constraints_ext.get('ca'):
+            values.append('CA:' + str(basic_constraints_ext['ca']).upper())
+
+        if basic_constraints_ext.get('path_length'):
+            values.append('pathlen:' + basic_constraints_ext['path_length'])
+
+        exts['basicConstraints'] = ', '.join(values)
```

Comment:
I am not sure about this implementation. If the 3 `if` conditions all evaluate to `False` then `exts['basicConstraints'] = ', '.join(values)` adds the empty string `""` into the `dict`. The same is true in the other commits, except the fourth, where something is unconditionally added to `exts`. Here we unconditionally add the key `basicConstraints`, even if there are no values.

My concern here is that doing something like...
```
>>> exts['basicConstraints'] is not None
True
```
...or...
```
>>> 'basicConstraints' in exts
True
```
...makes it look like there are values present when in reality there are none.

Evidently this is not a problem for the CI, as it passed, but I feel like we shouldn't be adding these empty objects in. If it is expected that if when `basic_constraints_ext` is `True` at least one of the following 3 conditions must be `True`, we should be raising some sort of validation error here. If it is expected that none of the 3 conditions need to be met then we should not add in the empty key and let the calling code raise `KeyError`. Are we checking for this empty string condition before doing anything with the extension later?

You didn't add this code, so I will approve, but should we address this?

Reply:
@ckelleyRH Thanks for the ACK. Let me try to clarify.

In the original Python code the `basicConstraints` property will not be added if `basic_constraints_ext` parameter is `None` or empty:
https://github.com/dogtagpki/pki/blob/master/base/common/python/pki/nssdb.py#L1182-L1195

In that case the `basicConstraints` variable in Java will be `null`, so the extension will not be added:
https://github.com/dogtagpki/pki/blob/master/base/common/src/main/java/org/dogtagpki/nss/NSSExtensionGenerator.java#L121-L122

If the `basic_constraints_ext` parameter in Python is not empty, there are some possibilities:

(a) `basic_constraints_ext` contains only valid values (i.e. `critical`, `ca`, `path_length`)
(b) `basic_constraints_ext` contains both valid and invalid values
(c) `basic_constraints_ext` contains only invalid values (due to typos or incorrect usage)

In case (a) the `basicConstraints` property in Python will have some values so the extension will be added in Java with to those values. However, not every value has to be specified since we have default values:
https://github.com/dogtagpki/pki/blob/master/base/common/src/main/java/org/dogtagpki/nss/NSSExtensionGenerator.java#L126-L128

In case (b) the Python code will drop the invalid values, so the extension will be added in Java with the valid values only, and possibly some default values too.

I think the scenario that you described is actually case (c). In this case the `basicConstraints` property in Python will be an empty string, so the extension will be added in Java with just the default values.

I'm not sure if it's valid to add the extension with just the default values, but that's not something we're trying to enforce here. If we want to enforce something it needs to be based on RFCs. Also, the entity using the cert is ultimately responsible to validate the extension.

The Java code actually will reject invalid values:
https://github.com/dogtagpki/pki/blob/master/base/common/src/main/java/org/dogtagpki/nss/NSSExtensionGenerator.java#L151
but since the Python code has already dropped the invalid values the Java code will never see them.

We could add more validations in the Python code, but that means we'll be duplicating the Java code. Also, considering this code so far is mostly used internally by our tools, I'm not sure it's worth the effort.

In my opinion case (c) is handled adequately for now since we have the default values, and the new code doesn't change that behavior.

Just a note, the current CI doesn't have much coverage on cert extensions so I wouldn't rely too much on it to assume that the current code is correct. We certainly need to add more positive tests for Java & Python API. We could add negative tests too to test the validations, but that's probably a lower priority.

Reply:
Thanks for the explanation! If there is already validation in the Java then it is OK, it is just nicer to prevent dubious input as close to the source as possible. Though as you say the use is internal, so it is not so critical.